### DISCUSSION TITLE
could we allow run replica clickhouse instance on same data directory ?

### DISCUSSION THREAD

**Discussion by qingant** [ClickHouse/ClickHouse#30956] [Ideas]
Currently multi instances running on same data directory is prevented by /data/status file lock. Well, it's reasonable to have a single replica instance for instant reporting query and one primary instance for heavy tasks like pre aggregations or clickhouse style streaming processing.

is there any essential difficulty to make this happen ?

**Comment by den-crane** [comment]
>Well, it's reasonable to have a single replica instance for instant reporting
> query and one primary instance for heavy tasks like pre aggregations or clickhouse style streaming processing.

It's not reasonable. There is no difference if you have 2 CH instances or 1 instance on the same node. 2 instances will be struggle for resources the same way as 1 internally.

>is there any essential difficulty to make this happen ?

Yes, the current architecture does not allow it. It's impossible. Use external/block storage HDFS / S3 and zero copy replication.

**Reply by qingant** [reply]
> > Well, it's reasonable to have a single replica instance for instant reporting
> > query and one primary instance for heavy tasks like pre aggregations or clickhouse style streaming processing.
> 
> It's not reasonable. There is no difference if you have 2 CH instances or 1 instance on the same node. 2 instances will be struggle for resources the same way as 1 internally.

hi @den-crane  there is difference at least on resource control, we can limit the one primary to use up to 70% of the resource so that the replica would always get the resource to response the query.
> 
> > is there any essential difficulty to make this happen ?
> 
> Yes, the current architecture does not allow it. It's impossible. Use external/block storage HDFS / S3 and zero copy replication.
1. if an instance never write, would it be possible ? and since there is some kind of high performance shared file system , we can actually run two instances in difference nodes.
